Sources/Kafka/RDKafka/RDKafkaUnsafeProducerMessage.swift

```diff
@@ -0,0 +1,112 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the swift-kafka-client open source project
+//
+// Copyright (c) 2023 Apple Inc. and the swift-kafka-client project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of swift-kafka-client project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Crdkafka
+// For `strdup` and `free`
+#if canImport(Glibc)
+import Glibc
+#else
+import Darwin.POSIX
+#endif
+
+/// Wrapper type for initializing a `librdkafka` producer message
+/// using the variadic argumetn initializer (required to set message headers).
+internal class RDKafkaUnsafeProducerMessage {
+    internal private(set) var _internal: [rd_kafka_vu_t]
+
+    internal let size: Int
+
+    init(
+        topicHandle: OpaquePointer,
+        partition: Int32,
+        messageFlags: Int32,
+        key: UnsafeRawBufferPointer?,
+        value: UnsafeRawBufferPointer,
+        opaque: UnsafeMutableRawPointer?,
+        headers: [KafkaHeader]
+    ) {
+        var index = 0
+        let sizeWithoutHeaders = (key != nil) ? 6 : 5
+        self.size = sizeWithoutHeaders + headers.count
+        self._internal = Array(repeating: rd_kafka_vu_t(), count: self.size)
+
+        self._internal[index].vtype = RD_KAFKA_VTYPE_RKT
+        self._internal[index].u.rkt = topicHandle
+        index += 1
+
+        self._internal[index].vtype = RD_KAFKA_VTYPE_PARTITION
+        self._internal[index].u.i32 = partition
+        index += 1
+
+        self._internal[index].vtype = RD_KAFKA_VTYPE_MSGFLAGS
+        self._internal[index].u.i = messageFlags
+        index += 1
+
+        if let key {
+            self._internal[index].vtype = RD_KAFKA_VTYPE_KEY
+            self._internal[index].u.mem.ptr = UnsafeMutableRawPointer(mutating: key.baseAddress)
+            self._internal[index].u.mem.size = key.count
+            index += 1
+        }
+
+        self._internal[index].vtype = RD_KAFKA_VTYPE_VALUE
+        self._internal[index].u.mem.ptr = UnsafeMutableRawPointer(mutating: value.baseAddress)
+        self._internal[index].u.mem.size = value.count
+        index += 1
+
+        self._internal[index].vtype = RD_KAFKA_VTYPE_OPAQUE
+        self._internal[index].u.ptr = opaque
+        index += 1
+
+        for header in headers {
+            self._internal[index].vtype = RD_KAFKA_VTYPE_HEADER
+
+            // Copy name C String (strdup) as we do not own the headers: [KafkaHeader] array
+            guard let headerNameCPointer = header.key.withCString(strdup) else {
```

Comment:
I am not really happy with doing the `strdup` here. However, we need to somehow need to pass a `char *` pointer which we get from the `headers: [KafkaHeader]` array (hence accessing the pointer inside of the closure is weird).

I had a couple of ideas to reduce copying here:

* write C glue code that takes the `headers: [KafkaHeader]` array and does essentially what `RDKakfaUnsafeProducerMessage` does, just in a more enclosed way
* use copy-on-write / a String pool to avoid unnecessary copies of the String pointer (if a message header name occurs more than once we can avoid the additional copy with our String pool)